### PR TITLE
Barlt flac comment extractor

### PR DESCRIPTION
**Pull request by noahmacaskill**
First draft of the FLAC comment extractor for BarLt. Doug took a test FLAC recording using firmware version 3.30, decided I'd create a fixture for it. It's got an interesting bug (which is what is causing the tests to fail), the SD card serial number from the log file doesn't match the serial from the CID in the comment block. A little more digging and I found the issue.

In the CID the serial number is given as an 8 bit long hex string, in our case 437AB370 (1132114800 when converted to decimal). The actual serial number (from the log file) is 8041328, or 7AB370 in hex. For some reason the CID isn't prefixing the hex string with 0s, but instead with seemingly random values. Maybe a bug with this firmware version? Any thoughts on how to tackle this?

### CODE REVIEW

File: src/MetadataUtility/Audio/Wave.cs
Comment (noahmacaskill) [inline src/MetadataUtility/Audio/Wave.cs:163]: In [this](https://github.com/QutEcoacoustics/emu/blob/master/test/Fixtures/FL_BAR_LT/3.14_PreallocatedHeader/20191125_AAO/20191125T000000%2B1000_REC.flac) preallocated header fixture, the data section isn't empty. Was looking for some other indicator I can use to classify it as being a faulty file, noticed that the riff chunk size is short of the actual size. Maybe it makes more sense to rename the function to "IsFaultyFile", since that may not necessarily be indicative of a preallocated header? What do you think?

**Comment by noahmacaskill** [comment]
Also removed [this](https://github.com/QutEcoacoustics/emu/blob/master/test/Fixtures/FL_BAR_LT/3.14_Normal/logfile.txt) log file (copied from glitchy signal fixture) that I added earlier, turns out I was wrong about that. The exact name of the recording is written in the log, but they are only named by date, is it possible that it was a different recording in the log, taken at the same time? I can tell they're different because the SD card info from the log doesn't match up with what I pulled out of the CID.

**Comment by noahmacaskill** [comment]
Besides that, welcoming any other feedback!

File: src/MetadataUtility/Audio/Wave.cs
Comment (atruskie) [reply src/MetadataUtility/Audio/Wave.cs:163]: So, for the fixture we have, it is an example of a _specific problem_ (in this case a pre-allocated file). You could make a `IsFaulyFile` method but it would probably invoke a few different detectors in the end since after all there are all sorts of problems. So yes, we want a specific check for pre-allocated header file. There are probably a few variants of the pre-allocated header problem.. i'm not sure at which stage FL overwrites the header with the actual flac header... we'll just have to evolve this as we go.

I think roughly the metrics for pre-allocated detection should be:

- it is a wave file with a .flac extension
- the riff chunk size is incorrect
- the data chunk size is incorrect
- first few samples of data are zeros (or there are no samples)
- the file is small (< 200 bytes)

And I think we should have this as a scoring system. If 3 out of 5 of any of these ring true then we classify it as a pre-allocated header.

Hopefully size of the file will be the biggest discriminator between pre-allocated header and a power failure (there are even more subtleties here we can get into later)

**Comment by atruskie** [comment]
Oh wow that's a ton on updates.

---

For the weird CID number file a report here please: https://github.com/ecoacoustics/known-problems
Gather evidence from a couple of files - you cited random behavior, so I'd need to spread of random values from files taken from the same deployment with the same SD card and sensor.

After you've filed the report, we'll contact Frontier Labs and see if they can offer an official statement and recommended fix.

---

> Also removed [this](https://github.com/QutEcoacoustics/emu/blob/master/test/Fixtures/FL_BAR_LT/3.14_Normal/logfile.txt) log file (copied from glitchy signal fixture) that I added earlier, turns out I was wrong about that. The exact name of the recording is written in the log, but they are only named by date, is it possible that it was a different recording in the log, taken at the same time? I can tell they're different because the SD card info from the log doesn't match up with what I pulled out of the CID.

I think I understand, nice, good detective work

**Review by atruskie — CHANGES_REQUESTED** [review]
Generally I'd like to keep the functions in the Flac and Wave classes vendor agnostic.

I nice case for this would be parsing the FL header comments from either Wave or Flac files:

- Flac has a read comments function that returns  dictionary<string, string>
- Wave has a read comments function that returns dictionary<string, string>
- FrontierLabs has a parse comments function that accepts the dictionary and converts the values 

Now, I'm not suggesting you implement the Wave aspect of this, just demonstrating the power of appropriate abstraction and compo sable functions.

(I think I may have made some misleading comments in the review (sorry, long review I got lost) about placements of things, but if you adhere to the above guidance it should work out)

File: src/MetadataUtility/Models/Sensor.cs
Comment (JordanMercerECCC) [reply src/MetadataUtility/Models/Sensor.cs:42]: I believe you are correct. The manual seems to suggest it.

File: src/MetadataUtility/Audio/Vendors/FrontierLabs.cs
Comment (noahmacaskill) [reply src/MetadataUtility/Audio/Vendors/FrontierLabs.cs:29]: Forgot to remove that. In the comment block each piece of microphone metadata is shown once for each channel (MicrophoneType1 and MicrophoneType2, etc.). When only one mic is hooked up, it displays "unknown" for each value except gain, where it is 0dB. Initially I was checking for this to mean non-existent mic, but am now just checking for a [UID of "unknown"](https://github.com/QutEcoacoustics/emu/blob/a8366ca4a1a10646f8a5d94a4bdb015c0e32a50e/src/MetadataUtility/Audio/Vendors/FrontierLabs.cs#L442)

File: src/MetadataUtility/Models/Sensor.cs
Comment (noahmacaskill) [reply src/MetadataUtility/Models/Sensor.cs:42]: Interesting, FLAC header has both ChannelGain1 and ChannelGain2

Comment (atruskie) [reply src/MetadataUtility/Models/Sensor.cs:42]: FL can store the values independently.

I'm not sure how we resolve this into the master gain setting for the recording, but for now, we should keep both (per-mic and general gain) settings.

Though thinking on it, perhaps this Gain should be moved to the `Sensor` record

File: src/MetadataUtility/Audio/Vendors/FrontierLabs.cs
Comment (atruskie) [reply src/MetadataUtility/Audio/Vendors/FrontierLabs.cs:10]: Ok, Models shouldn't be imported into this class so that's one fix.

But given the use of LanguageExt in this class I'd guess it'd be more prevalent than the models

Comment (noahmacaskill) [reply src/MetadataUtility/Audio/Vendors/FrontierLabs.cs:10]: Makes sense, won't be needed in here once I rearchitect.

**Comment by noahmacaskill** [comment]
Almost done rearchitecting this one. Just a note on the weird CID bug. [I made a new issue in known problems](https://github.com/ecoacoustics/known-problems/issues/1), this was just from a test recording Doug did at home so we don't have much of any data on how common the issue is, other than doing a second test recording (with a different SD card) and having the same problem. Haven't thought of a way to flag the problem from a software perspective either since the CID otherwise appears normal.

**Comment by noahmacaskill** [comment]
I added a little more context and an update over on the known-problems repo. I could replace the most significant byte with 00 to make the values match up, but I could potentially then be overwriting actual serial number values (if a serial number were large enough to need all 4 bytes). I'm also now not sure if the value from the log file, CID, or either is the correct one.

**Comment by atruskie** [comment]
Yeah so something I don't think I've articulated is I originally envisioned keeping track of the source for each bit of metadata as well. Then in cases where multiple bit of metadata for the same field were captured we could make a choice over which is more accurate/complete bit of data and smartly choose whether or not to overwrite the missing data.

This is not an idea that made it into the current iteration of the metadata extractor.

However we can special case for this situation and I we should:

- run the header extractor first
- run the log file extractor second
   - the log file extractor should not overwrite information already obtained
   - if we detect this CID issue then you choose which bit of metadata to keep

**Comment by noahmacaskill** [comment]
Think I've made some good progress here. Have a few questions/clarifications.

**Comment by noahmacaskill** [comment]
> Yeah so something I don't think I've articulated is I originally envisioned keeping track of the source for each bit of metadata as well. Then in cases where multiple bit of metadata for the same field were captured we could make a choice over which is more accurate/complete bit of data and smartly choose whether or not to overwrite the missing data.
> 
> This is not an idea that made it into the current iteration of the metadata extractor.
> 
> However we can special case for this situation and I we should:
> 
> * run the header extractor first
> * run the log file extractor second
>   
>   * the log file extractor should not overwrite information already obtained
>   * if we detect this CID issue then you choose which bit of metadata to keep

Makes sense, I've made it so the header extractor runs first, and once data is written by an extractor, it won't be overwritten. I'm still not sure which serial number is the correct one, seems like you think it's the one written by the CID? If that's the case, then what I've done should fix the problem? The serial number from the CID is what's now showing up in the final output.

**Comment by noahmacaskill** [comment]
You mentioned reading in each each block from FLAC files one by one like we do with Wave (using ranges). As of now I'm finding the vorbis comment block [here](https://github.com/QutEcoacoustics/emu/blob/e559b801eec061bdc01eb5941268d5192cd49447/src/MetadataUtility/Audio/Flac.cs#L258-L283). My FLAC Header extractor is just using absolute offsets to parse each value since the places are guaranteed for FLAC.

Do you think it makes sense to refactor this to work more like Wave? I could first read in the fLaC magic number. Then the Stream Info block (could adjust my offsets from absolute to relative to the stream info block to extract those values). Then write a function similar to [this](https://github.com/QutEcoacoustics/emu/blob/e559b801eec061bdc01eb5941268d5192cd49447/src/MetadataUtility/Audio/Wave.cs#L311-L365) to scan for the vorbis comment block?

**Comment by atruskie** [comment]
> I'm still not sure which serial number is the correct one, seems like you think it's the one written by the CID? 

I finally dedicated some brain power to this and I'm pretty sure the serial inside the CID is correct. I imagine you've read it already but it would be great for you to double check our logic: https://github.com/ecoacoustics/known-problems/blob/main/frontier_labs/FL007.md

> If that's the case, then what I've done should fix the problem? The serial number from the CID is what's now showing up in the final output.

Great.

> That leaves figuring out how to get the log file test to pass. Right now it's failing since the serial number in the Fixtures csv is the serial number from the CID. Thought of adding a "Problems" column to the csv, where I can add something like "SD Card Serial Number" to my [new fixture](https://github.com/QutEcoacoustics/emu/blob/barlt-flac-comment-extractor/test/Fixtures/FL_BAR_LT/3.30_NewFirmware/00008570_Dartmouth_20220331T094900-0300_Rec.flac), then in the log file extractor test I'll skip the serial number assertion? Let me know if you have any better ideas for that.

We've got to be careful how much we overload that Fixtures spreadsheet - it already is a bit unmaintainable. 

If you did add problems column we'd encode the problem with the unique identifier we assigned if in the known-problems document (_FL007_).

I think the best course of action is for the LogFileExtractor to simply not report the serial if FL007 is detected (null for  the serial field). 

Then we need some way to discriminate in tests:

  - either duplicate the fixture row, one row would have the extractor column include the header extractor and one would have the log file extractor set; then for the problem files of FL007 the serial column would be null
      - we could extend this method too eventually with a combined row that tests what happens when both extractors are active
  - we could hard code something into the test of the log file extractor, something like:
    ```
    if (target affected by FL007) {
       expected.serial = null; // it's not possible to read an accurate serial number in this case
    }
    ```

And all of this is more evidence I need to rework the fixtures csv. I think YAML might be appropriate - we can use YAML references to repeat blocks of repeated information and then for each extractor combination override bits of information succinctly that pertain to the test at hand. But that is a task for another time!

> You mentioned reading in each each block from FLAC files one by one like we do with Wave (using ranges). As of now I'm finding the vorbis comment block [here](https://github.com/QutEcoacoustics/emu/blob/e559b801eec061bdc01eb5941268d5192cd49447/src/MetadataUtility/Audio/Flac.cs#L258-L283). My FLAC Header extractor is just using absolute offsets to parse each value since the places are guaranteed for FLAC.

As I understood the FLAC spec, the STREAMINFO  must come after the magic `fLaC` marker. After that however each metadata block could arrive in any order. So the extractor that works streaminfo can continue to work off fixed offsets. It looks like the code you linked to is doing the correct thing for the comment block finding.

> Do you think it makes sense to refactor [this](https://github.com/QutEcoacoustics/emu/blob/e559b801eec061bdc01eb5941268d5192cd49447/src/MetadataUtility/Audio/Wave.cs#L311-L365) to work more like Wave? I could first read in the fLaC magic number. Then the Stream Info block (could adjust my offsets from absolute to relative to the stream info block to extract those values). Then write a function similar to this to scan for the vorbis comment block?

Yes I think you decouple the find block algorithm from the FindVendorString function. Actually I'm not sure what the vendor string function does at all?

More comments in review...

File: src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs
Comment (atruskie) [inline src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs:96]: OK this is not leaky (good) but it's a right mess (not your fault).

I would suggest breaking this up a bit by pulling common expressions our into variables.

File: src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs
Comment (atruskie) [inline src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs:76]: no need to limit this to just two microphones

File: src/MetadataUtility/Models/Microphone.cs
Comment (atruskie) [inline src/MetadataUtility/Models/Microphone.cs]: Channel should be present, why did you remove it?

**Review by atruskie — CHANGES_REQUESTED** [review]
Ok lots of comments, we'll talk in more detail soon.

But on the whole a great improvement! Lots of tricky details to consider

File: src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs
Comment (noahmacaskill) [inline src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs:91]: I think this is a big improvement. I've created a method between the recording and the comment parsing (ParseComment) to do two checks for us.

1. Make sure the comment exists
2. Make sure there are no errors while parsing (check the fin), logging them if there are

Let me know what you think.

**Comment by noahmacaskill** [comment]
I've also added a new row to our fixtures files containing the SD card serial number from the log file like we discussed last week. This has led to a few more tests failing. Some of our tests are tied to the "Process" column in our fixtures csv. For example, the FlacCommentExtractorTests [CanProcessFilesWorks function](https://github.com/QutEcoacoustics/emu/blob/9134f2cf9a69c2e590bcf22e530ef25112557923/test/MetadataUtility.Tests/Metadata/FrontierLabs/FlacCommentExtractorTests.cs#L28-L37) uses it to determine whether the CanProcess function for that extractor should pass or fail. 

The CanProcess function returns true for our new fixture (since there is a vorbis comment block in the file), but the expected value will be false (since we've removed that extractor from the Process column).

One idea I had was to add a "Problems" column to the CSV (which would have FL007 for our new fixture). Then when testing our CanProcess function we do something like:

`var expected = model.Process.Contains("FlacCommentExtractor") || model.Problems.Contains("FL007")`

We could potentially refactor some other tests to use this column as well with other known faulty fixtures files. Not ideal to keep adding to the already bulky CSV though, what do you think?

File: src/MetadataUtility/Models/Microphone.cs
Comment (noahmacaskill) [reply src/MetadataUtility/Models/Microphone.cs]: I originally used this to keep track of which microphone data belong to which microphone, pulling the channel number from the microphone vorbis comments (ex. MicrophoneTypeX where X is channel). I wasn't sure if this actually was the channel number though, so I removed it. Looking at the log files, they're referred to as microphone channel A and channel B. Could bring them back using the values from the vorbis comments (as integers?), or convert to alphabetical?

Comment (atruskie) [reply src/MetadataUtility/Models/Microphone.cs]: Because this structure is essentially removing the order we should preserver the order in which the information was conveyed to us.

Channel should be a index (A = 0, and so on)
Maybe add a ChannelName property as well so we don't lose the `A`/`B` part of the information.

**Comment by atruskie** [comment]
> `tExtractor") || model.Problems.Contains("FL007")`
> 
> We could potentially refactor some other tests to use this column as well with other known faulty fixtures files. Not ideal to keep adding to the already bulky CSV though, what do you think?

Another idea is to split the `Process` column into `Process` and `CanProcess`?

I think testing for a specific edge case in the test is just putting us in an echo chamber (since we use the same test in the CanProcess function itself).

Regardless we don't need to over engineer it now, since I'm convinced I need to change the data format of the fixtures CSV anyway

File: src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs
Comment (atruskie) [reply src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs:91]: much cleaner, yeah

**Review by atruskie — APPROVED** [review]
OK, changes look good.

Once tests are passing we should be good to merge

**Comment by noahmacaskill** [comment]
Aside from that I fixed up the tests and patched up a couple small things. Good to merge if everything looks good!

File: src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs
Comment (atruskie) [reply src/MetadataUtility/Metadata/FrontierLabs/FlacCommentExtractor.cs:50]: > Converted channel number to channel name as its given in the log file like we discussed.

> Great

**Comment by atruskie** [comment]
@noahmacaskill - I forgot to get you to update the docs/metadata.ms document with the new fields that can be extracted. Can you please do that?

**Pull request merged**
